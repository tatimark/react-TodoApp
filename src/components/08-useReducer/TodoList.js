import React from 'react';
import './styles.css';
import { TodoListItem } from './TodoListItem';

export const TodoList = ({todos, handleDelete, handleToggle}) => {
    return (
      
            <ul className="list-group list-group-flush">
        {
            todos.map( (todo, i ) => (
                // TodoListItem, todo, index, handleDelate, handleToggle
                <TodoListItem 
                 key={todo.id}
                 todo={todo}
                 index={i}
                 handleDelete={handleDelete} 
                 handleToggle={handleToggle}
                />
            ))
        }
        </ul> 
       
    )
}


