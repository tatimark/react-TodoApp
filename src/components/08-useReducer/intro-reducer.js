

const inicialState = [{
    id: 1,
    todo: 'Comprar leche',
    done: false
}];

const todoReducer = ( state = inicialState , action ) =>{
    if(action?.type === 'agregar'){
        return [...state, action.payload];
    }
    return state;
}


let todos = todoReducer();

// no utilizar el push 


const newTodo = {
    id: 2,
    todo: 'Comprar leche',
    done: false
}
// que tipo de acci√≥n es al reducer
const agregarTodo = {
    type: 'agregar',
    payload: newTodo
}

todos = todoReducer(todos, agregarTodo);
