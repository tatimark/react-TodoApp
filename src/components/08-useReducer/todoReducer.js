export const todoReducer = (state = [], action) => {

        switch (action.type) {
            case 'add':

                return [...state, action.payload];

            case 'delete':
                return state.filter(todo => todo.id !== action.payload); //payload 123535 action.payload.id regresa un nuevo arreglo


            case 'toggle':
                    return state.map( todo =>
                        (todo.id === action.payload) //ternario
                        ?{...todo, done: !todo.done }
                        : todo
                    );

            case 'toggle-old':
                return state.map(todo => {
                    if (todo.id === action.payload) {
                        return {
                            ...todo,
                            done: !todo.done
                        }
                    } else {
                        return todo; // hay que regresar algo para que no haya undefind
                    }

                })

            default:
                return state;
        }
    }