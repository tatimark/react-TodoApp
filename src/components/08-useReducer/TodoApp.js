import React, { useReducer } from 'react';
import { useEffect } from 'react';

import './styles.css';
import { todoReducer } from './todoReducer';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';

// const initialState = [{
//     id: new Date().getTime(),
//     desc: 'Aprender React',
//     done: false
// }]

const init= () =>{
    // si hay todos traelos

    return JSON.parse(localStorage.getItem('todos')) || []; // retorna todos si hay o sino un arreglo vacio

    // return localStorage.getItem('todos');
    // return [{
    //     id: new Date().getTime(),
    //     desc: 'Aprender React',
    //     done: false
    // }]
}

export const TodoApp = () => {

const [ todos , dispatch ] = useReducer(todoReducer, [], init); // el InitialState es el init y [] es lo que va a ser inicialmente





useEffect( () =>{ 
    localStorage.setItem('todos', JSON.stringify( todos )); // solo guarda string el localstorage
}, [todos] )


const handleDelete = (todoId) =>{
    console.log(todoId);
    // crear la action

    // const newTodoDelete = { asÃ­ funciona
    //     id: todoId,
    //     desc: description,
    //     done: false
    //    };
    //    const actionDelete = {
    //        type: 'delete',
    //        payload: newTodoDelete
    //    }

    const action = {
           type: 'delete',
            payload: todoId
    }
    

    //dispatch
    dispatch(action);
//     dispatch(actionDelete);
 }


 const handleToggle = ( todoId ) =>{
     dispatch({
         type: 'toggle',
         payload: todoId
     })

 }

const handleAddTodo = ( newTodo ) =>{

    dispatch({
        type: 'add',
        payload: newTodo
    });
}

// console.log(description);

// const handleSubmit = (e) => {
//     e.preventDefault();

//     if( description.trim().length <= 1){
//         return;
//     }
//    const newTodo = {
//     id: new Date().getTime(),
//     desc: description,
//     done: false
//    };
  
//    reset();

// }

    return (
        <div>
            <h1>TodoApp ( { todos.length } )</h1>
            <hr />
            <div className="row">
                <div className="col-7">
              
        {/* TodoList,  necesita los todos argumento enviado mediante prop, handleDelete, handleToggle */}
                    <TodoList 
                    todos={ todos}
                    handleDelete={ handleDelete} 
                    handleToggle={ handleToggle}
                    />
                        {/* <ul className="list-group list-group-flush">
                            {
                                todos.map( (todo, i ) => (
                                    // TodoListItem, todo, index, handleDelate, handleToggle
                                <li
                                    key = { todo.id }
                                    className="list-group-item"
                                >
                                    <p className={ `${ todo.done && 'complete' }`}
                                    onClick={ () => handleToggle(todo.id)}
                                    >{ i + 1 }. { todo.desc }</p>
                                    <button
                                    className="btn btn-danger"
                                    onClick={ () => handleDelete(todo.id) }
                                    >
                                        Borrar
                                    </button>
                                    </li>
                                ))
                            }
                            </ul>  */}
                </div>
                <div className="col-5">
                    <TodoAdd 
                        handleAddTodo={ handleAddTodo }
                    />
                    {/* <h4>Agregar Todo</h4>
                    <hr />
                    <form onSubmit={ handleSubmit }>
                    <input 
                        type="text"
                        name="description"
                        className="form-control"
                        placeholder="Aprender ..."
                        autoComplete="off"
                        value={ description }
                        onChange={ handleInputChange }
                    />
                    
                    <button
                        type="submit"
                        className="btn btn-outline-primary mt-1 btn-block"
                    >
                        Agregar
                    </button>
                    </form> */}
                </div>
            </div>
     
        </div>
    )
}
