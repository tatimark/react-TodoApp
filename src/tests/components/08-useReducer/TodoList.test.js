import React from 'react';
import { shallow } from 'enzyme';
import { TodoList } from '../../../components/08-useReducer/TodoList';
import { demoTodos } from '../fixtures/demoTodos';

describe(' Prueba en Todolist', () => {

    const handleDelete = jest.fn();
    const handleToggle = jest.fn();

    const wrapper = shallow(
        <TodoList
        todos={ demoTodos }
        handleDelete={ handleDelete }
        hangleToggle={ handleToggle }
        />
    )

    test('debe de mostrarse correctamente ', () => {
       
        expect( wrapper ).toMatchSnapshot();
    })

    test('debe de tener dos TodoListItem', () => {

        expect(wrapper.find('TodoListItem').length).toBe( demoTodos.length);


        console.log(wrapper.find('TodoListItem').at(0).props());
       expect(wrapper.find('TodoListItem').at(0).prop('handleDelete')).toEqual( expect.any(Function)); //espera cualquier funci√≥n
       

    })


})