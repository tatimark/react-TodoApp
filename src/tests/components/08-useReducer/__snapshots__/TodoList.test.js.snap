// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` Prueba en Todolist debe de mostrarse correctamente  1`] = `
<ul
  className="list-group list-group-flush"
>
  <TodoListItem
    handleDelete={[MockFunction]}
    index={0}
    key="1"
    todo={
      Object {
        "desc": "Aprender React",
        "donde": false,
        "id": 1,
      }
    }
  />
  <TodoListItem
    handleDelete={[MockFunction]}
    index={1}
    key="2"
    todo={
      Object {
        "desc": "Aprender Mongo",
        "done": false,
        "id": 2,
      }
    }
  />
</ul>
`;
